src/database-controller/src/common/framework.js

```diff
@@ -290,6 +333,20 @@ class Snapshot {
     }
     this._snapshot = jsonmergepatch.apply(this._snapshot, patchData);
   }
+
+  setFailed() {
+    // Manually set the framework's status to failed.
+    // This is used when we cannot sync the framework request to api server,
+    // and we're sure that the error is unrecoverable.
+
+    // For now, we only use it for frameworks that never start.
+    console.log(this.getTotalRetriedCount(), this.getState())
+    if (!this.getTotalRetriedCount() === 0 || this.getState() !== 'AttemptCreationPending') {
+      throw new Error('setFailed() only works for framework that never start!')
```

Comment:
and not synced to api server?

Reply:
`requestSyneced` is not managed by snapshot class.

For this change, we are sure that the framework's `requestSynced` field is false because we call `setFailed()` from `synchronizeHandler`. We only call `synchronizeHandler` for  `request=false` frameworks.

Reply:
So add a comment here

Reply:
done